# Tugas dan praktikum Naive Bayes

In [3]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)


     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


# Bagian 1: Import library dan memuat dataset

#### Di bagian awal, kita memanggil tiga alat yang akan dipakai sepanjang program. NumPy dipakai untuk mengolah angka dalam bentuk array, Pandas dipakai untuk membuat tabel data, dan load_iris adalah fungsi dari scikit-learn untuk mengambil dataset Iris yang sudah tersedia. Setelah iris = load_iris() dijalankan, variabel iris berisi seluruh data: 150 bunga iris, masing-masing punya empat ukuran (panjang dan lebar kelopak, panjang dan lebar mahkota), beserta jenis spesiesnya.

In [4]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

# Bagian 2: Mengubah data menjadi DataFrame
Data mentah dari scikit-learn berbentuk array yang sulit dibaca, jadi kita ubah menjadi tabel dengan pd.DataFrame, lengkap dengan nama kolomnya. Kemudian kita tambahkan kolom Species yang berisi nama spesies (setosa, versicolor, virginica), bukan lagi angka 0, 1, dan 2. Setelah tabel siap, kita pisahkan dua hal. X adalah semua kolom ukuran bunga yang menjadi bahan bagi model untuk belajar, sedangkan y adalah kolom Species yang menjadi jawaban yang ingin ditebak oleh model. Perintah print(X) hanya untuk melihat isi tabel ukurannya.

In [6]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

# Bagian 3: Mengubah label teks menjadi angka
Model machine learning tidak bisa memproses tulisan seperti "setosa", jadi nama spesies perlu diganti dengan angka. LabelEncoder mengerjakan ini secara otomatis dengan mengurutkan nama spesies sesuai abjad, sehingga setosa menjadi 0, versicolor menjadi 1, dan virginica menjadi 2. Objek lab sengaja kita simpan karena nanti dibutuhkan untuk menerjemahkan angka hasil prediksi kembali menjadi nama spesies.

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

# Bagian 4: Membagi data latih dan data uji
Sebelum model dilatih, data dibagi dua. Sebanyak 80 persen (120 bunga) dipakai untuk melatih model, dan 20 persen sisanya (30 bunga) disimpan untuk menguji seberapa pintar model tersebut. Pembagian ini penting karena kita ingin menguji model dengan data yang belum pernah dilihatnya, seperti ujian yang soalnya berbeda dari latihan. Pengaturan random_state=0 membuat pengacakan datanya selalu sama setiap kali program dijalankan, sehingga hasilnya konsisten. Pembagian dilakukan dua kali, satu untuk ukuran bunga dan satu untuk labelnya, dan keduanya tetap cocok karena memakai pengaturan acak yang sama.

In [8]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)
result = NB_model.predict(X_test)
result


array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

# Bagian 5: Melatih model Gaussian Naïve Bayes
Di sini model dibuat dan dilatih. Kita memilih Gaussian Naïve Bayes karena semua ukuran bunga berupa angka kontinu yang bisa dianggap menyebar mengikuti kurva normal. Saat fit dijalankan, model mempelajari data latih dengan cara menghitung rata-rata dan sebaran ukuran bunga untuk setiap spesies. Misalnya, model belajar bahwa bunga setosa umumnya punya mahkota yang sangat kecil, sedangkan virginica punya mahkota yang besar. Model juga mencatat seberapa sering tiap spesies muncul di data latih.

In [9]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


# Bagian 6: Menguji model dan menghitung akurasi
Setelah belajar, model diminta menebak spesies dari 30 bunga di data uji lewat predict. Untuk tiap bunga, model membandingkan ukuran bunga itu dengan pola yang sudah dipelajari, lalu memilih spesies yang paling masuk akal. Hasil tebakan kemudian dibandingkan dengan jawaban yang sebenarnya memakai accuracy_score. Hasilnya 96,67 persen, artinya dari 30 bunga, 29 ditebak dengan benar dan hanya 1 yang salah.

In [11]:
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: [0]


C:\Users\MSII\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [12]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: [0]


# Tugas

In [13]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]),
                          columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: [2]


# Memprediksi data baru

Bagian terakhir adalah inti dari soal tugas nomor 2. Kita menyiapkan satu bunga baru dengan ukuran 7,3; 3,8; 4,6; dan 2,3, lalu membungkusnya menjadi tabel dengan nama kolom yang sama seperti data latih. Nama kolom ini penting supaya tidak muncul peringatan bahwa data tidak punya nama fitur. Model kemudian menebak spesies bunga tersebut dan menghasilkan angka 2. Karena angka itu sulit dipahami, inverse_transform dari lab menerjemahkannya kembali menjadi nama spesies, yaitu virginica. Tebakan ini masuk akal karena lebar mahkota 2,3 cm termasuk besar dan sangat khas untuk virginica.